zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/HDFSNotebookRepo.java

```diff
@@ -0,0 +1,241 @@
+    } catch (Exception e) {
+      throw new IOException(e.getCause());
+    }
+  }
+
+  private void syncHDFSNoteList() throws IOException {
+    FileObject rootDir = super.getRootDir();
+    FileObject[] children = rootDir.getChildren();
+    List<String> hdfsNotebook = getHdfsNotebook();
+
+
+    for (FileObject f : children) {
+      String baseName = f.getName().getBaseName();
+      logger.debug("read a notebook from local storage: " + baseName);
+
+      if (f.isHidden() || f.getType() != FileType.FOLDER ||
+         baseName.startsWith(".") || baseName.startsWith("#") || baseName.startsWith("~")) {
+        continue;
+      }
+
+      if (hdfsNotebook.contains(baseName)) {
+        hdfsNotebook.remove(baseName);
+      }
+      else {
+        uploadNoteListToHDFS(baseName);
+      }
+    }
+
+    for (String noteId : hdfsNotebook) {
+      downloadNotebook(noteId, rootDir);
+    }
+  }
+
+  private void uploadNoteListToHDFS(Note note) throws IOException {
+    uploadNoteListToHDFS(note.id());
+  }
+
+  private void uploadNoteListToHDFS(String noteId) throws IOException {
+    String localNotebook = super.getRootDir() + "/" + noteId + "/" + NOTE_JSON;
+    FileObject localNote = super.getRootDir().resolveFile(noteId + "/" + NOTE_JSON);
+    String noteDir = this.hdfsNotebookDir + "/" + noteId;
+    String notebook = noteDir + "/" + NOTE_JSON;
+    String newNotebook = noteDir + "/" + NOTE_JSON_TEMP;
+    logger.debug("localNotebook: {}\tnotebook: {}", localNotebook, notebook);
+
+    try {
+      this.hdfsCmd.runCommand(this.hdfsCmd.makeDirectory, noteDir, null);
+      this.hdfsCmd.runCommand(this.hdfsCmd.CreateWriteFile, newNotebook, localNote, null);
+      this.hdfsCmd.runCommand(this.hdfsCmd.DeleteFile, notebook, null);
+      Arg dest = this.hdfsCmd.new Arg("destination", notebook);
+      Arg[] renameArgs = {dest};
+      this.hdfsCmd.runCommand(this.hdfsCmd.RenameFile, newNotebook, renameArgs);
+    } catch (Exception e) {
+      logger.error("Exception: ", e);
+      throw new IOException(e.getCause());
+    }
+  }
+
+  private void removeHDFSNote(String noteId) throws IOException {
+    String noteDir = this.hdfsNotebookDir + "/" + noteId;
+    logger.debug("remove noteDir: " + noteDir);
+
+    Arg recursive = this.hdfsCmd.new Arg("recursive", "true");
+    Arg[] args = {recursive};
+
+    try {
+      this.hdfsCmd.runCommand(this.hdfsCmd.DeleteFile, noteDir, args);
+    } catch (Exception e) {
+      logger.error("Exception: ", e);
+      throw new IOException(e.getCause());
+    }
+  }
+
+  @Override
+  public List<NoteInfo> list(AuthenticationInfo subject) throws IOException {
+    if (this.enableWebHDFS)
+      syncHDFSNoteList();
+    return super.list(subject);
+  }
+
+  @Override
+  public synchronized void save(Note note, AuthenticationInfo subject) throws IOException {
+    super.save(note, subject);
+    if (this.enableWebHDFS)
+      uploadNoteListToHDFS(note);
```

Comment:
do you think about: 
  * whether makes sense to pass `user` from `subject` instead of getting user from environment inside `HDFSCommand`
  * from this function name seems like uploading whole list instead of a single note